src/expr.rs

```diff
@@ -1411,6 +1411,38 @@ fn rewrite_match_arm(
     )
 }
 
+/// Returns true if the given pattern is short. A short pattern is defined by the following grammer:
+///
+/// [small, ntp]:
+///     - single token
+///     - `&[single-line, ntp]`
+///
+/// [small]:
+///     - `[small, ntp]`
+///     - unary tuple constructor `([small, ntp])`
+///     - `&[small]`
+fn is_short_pattern(pat: &ast::Pat, pat_str: &str) -> bool {
+    // We also require that the pattern is reasonably 'small' with its literal width.
+    pat_str.len() <= 20 && !pat_str.contains("\n") && is_short_pattern_inner(pat)
```

Comment:
Do you think it is necessary to have the length check?

Reply:
I added the length check to work around some extreme cases:

```diff
Mismatch at tests/target/configs-match_pattern_separator_break_point-Front.rs:3:
 ⏎
 fn main() {⏎
     match lorem {⏎
-        aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa⏎
-        | bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb⏎                                                                              
-        | cccccccccccccccccccccccccccccccccccccccc⏎                                                                              
-        | dddddddddddddddddddddddddddddddddddddddd => (),⏎                                                                       
+        aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa | bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb⏎
+        | cccccccccccccccccccccccccccccccccccccccc | dddddddddddddddddddddddddddddddddddddddd => (),⏎                            
         _ => (),⏎
     }⏎
 }⏎

Mismatch at tests/source/configs-match_pattern_separator_break_point-Front.rs:3:
 ⏎
 fn main() {⏎
     match lorem {⏎
-        aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa⏎
-        | bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb⏎                                                                              
-        | cccccccccccccccccccccccccccccccccccccccc⏎                                                                              
-        | dddddddddddddddddddddddddddddddddddddddd => (),⏎                                                                       
+        aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa | bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb⏎
+        | cccccccccccccccccccccccccccccccccccccccc | dddddddddddddddddddddddddddddddddddddddd => (),⏎                            
         _ => (),⏎
     }⏎
 }⏎

Mismatch at tests/source/match.rs:19:
         Pattern1 | Pattern2 | Pattern3 => false,⏎
         Paternnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnn⏎
         | Paternnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnn => blah,⏎
-        Patternnnnnnnnnnnnnnnnnnn⏎
-        | Patternnnnnnnnnnnnnnnnnnn⏎                                                                                             
-        | Patternnnnnnnnnnnnnnnnnnn⏎                                                                                             
+        Patternnnnnnnnnnnnnnnnnnn | Patternnnnnnnnnnnnnnnnnnn | Patternnnnnnnnnnnnnnnnnnn⏎
         | Patternnnnnnnnnnnnnnnnnnn => meh,⏎
 ⏎
         Patternnnnnnnnnnnnnnnnnnn | Patternnnnnnnnnnnnnnnnnnn if looooooooooooooooooong_guard => {⏎

Mismatch at tests/target/match.rs:19:
         Pattern1 | Pattern2 | Pattern3 => false,⏎
         Paternnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnn⏎
         | Paternnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnn => blah,⏎
-        Patternnnnnnnnnnnnnnnnnnn⏎
-        | Patternnnnnnnnnnnnnnnnnnn⏎                                                                                             
-        | Patternnnnnnnnnnnnnnnnnnn⏎                                                                                             
+        Patternnnnnnnnnnnnnnnnnnn | Patternnnnnnnnnnnnnnnnnnn | Patternnnnnnnnnnnnnnnnnnn⏎
         | Patternnnnnnnnnnnnnnnnnnn => meh,⏎
 ⏎
         Patternnnnnnnnnnnnnnnnnnn | Patternnnnnnnnnnnnnnnnnnn if looooooooooooooooooong_guard => {⏎
test self_tests ... ok

Mismatch at tests/source/expr.rs:402:
     ];⏎
 ⏎
     match x {⏎
-        xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx⏎
-        | yyyyyyyyyyyyyyyyyyyyyyyyyyyyyy⏎                                                                                        
+        xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx | yyyyyyyyyyyyyyyyyyyyyyyyyyyyyy⏎
         | zzzzzzzzzzzzzzzzzzzzzzzzzzzzzz => foo(a, b, c),⏎
         _ => bar(),⏎
     };⏎

Mismatch at tests/target/expr.rs:402:
     ];⏎
 ⏎
     match x {⏎
-        xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx⏎
-        | yyyyyyyyyyyyyyyyyyyyyyyyyyyyyy⏎
+        xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx | yyyyyyyyyyyyyyyyyyyyyyyyyyyyyy⏎
         | zzzzzzzzzzzzzzzzzzzzzzzzzzzzzz => foo(a, b, c),⏎
         _ => bar(),⏎
     };⏎
```

Reply:
All these examples look like they should be caught by the `!pat_str.contains("\n")` clause, am I misunderstanding how that works?

Reply:
Sorry for ambiguity. The examples show the diffs when we remove `pat_str.len() <= 20` from the PR. 

Without the length check, we use mixed layout (`+` in diffs). With the length check, we use vertical layout (`-` in diffs). 

My first idea was to use vertical layout for excessively long patterns.

Reply:
AH, I see, thanks for explaining!